chainer/functions/connection/multihead_attention.py

```diff
@@ -0,0 +1,298 @@
+
+    Returns:
+        tuple: This function returns a tuple containing ``attn_output`` and
+        ``attn_output_weights``.
+
+        - ``attn_output`` is the output of attention whose shape is
+          :math:`(L, B, E)` where :math:`L` is the target sequence length,
+          :math:`B` is the batch size, and :math:`E` is the embedding size.
+        - ``attn_output_weights`` is the weights of attention whose shape is
+          :math:`(B, L, S)` where :math:`B` is the batch size,
+          :math:`L` is the target sequence length,
+          and :math:`S` is the source sequence length. If ``return_weights`` is
+          ``False``, this return value is ``None``.
+
+    .. seealso:: :class:`~chainer.links.MultiHeadAttention`
+
+    """
+    chainer.utils.experimental('chainer.functions.multihead_attention')
+
+    are_different = isinstance(proj_in_W, tuple)
+
+    def _in_proj(x, start=0, end=None, weight_idx=None):
+        if are_different:
+            W = proj_in_W[weight_idx]
+        else:
+            W = proj_in_W[start:end, :]
+        if proj_in_b is not None:
+            b = proj_in_b[start:end]
+        return linear.linear(x, W, b, n_batch_axes=x.ndim-1)
+
+    def _in_proj_qkv(query):
+        return split_axis.split_axis(_in_proj(query), 3, axis=-1)
+
+    def _in_proj_kv(key):
+        return split_axis.split_axis(
+            _in_proj(key, embedding_size), 2, axis=-1)
+
+    def _in_proj_q(query):
+        return _in_proj(query, end=embedding_size, weight_idx=0)
+
+    def _in_proj_k(key):
+        return _in_proj(
+            value, embedding_size, 2 * embedding_size, weight_idx=1)
+
+    def _in_proj_v(value):
+        return _in_proj(value, start=2 * embedding_size, weight_idx=2)
+
+    if embedding_size % n_head != 0:
+        raise ValueError(
+            '`embedding_size` ({}) need to be '.format(embedding_size) +
+            'divisible by `n_head` ({})'.format(embedding_size, n_head))
+    if (bias_k is None) != (bias_v is None):
+        raise ValueError
+    qkv_same = (query is key) and (query is value)
+    kv_same = key is value
+    target_length, batch_size, embedding_size = query.shape
+    head_size = embedding_size // n_head
+    if dot_product_scaler is None:
+        dot_product_scaler = head_size ** -0.5
+
+    device = chainer.backend.get_device_from_array(query)
+    xp = device.xp
+    dtype = query.dtype
+
+    if qkv_same:
+        # self-attention
+        q, k, v = _in_proj_qkv(query)
+    elif kv_same:
+        q = _in_proj_q(query)
+        if key is None:
+            k, v = None, None
+        else:
+            k, v = _in_proj_kv(key)
+    else:
+        q = _in_proj_q(query)
+        k = _in_proj_k(key)
+        v = _in_proj_v(value)
+    q *= dot_product_scaler
+
+    if bias_k is not None:
+        k = concat.concat((k, repeat.repeat(bias_k, batch_size, axis=1)))
+        v = concat.concat((v, repeat.repeat(bias_v, batch_size, axis=1)))
```

Comment:
I'm referencing https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L3294-L3297.

Will recheck this, including the difference between the behavior of `chainer.functions.repeat` and that of `torch.Tensor.repeat`.

Reply:
> These concat-s should be done in axis=0 rather than =1 (default)?

Looks it should. ([`chainer.functions.concat`](https://docs.chainer.org/en/stable/reference/generated/chainer.functions.concat.html) -> axis=1, [`torch.cat`](https://pytorch.org/docs/stable/torch.html#torch.cat) -> axis=0.)